heat/core/manipulations.py

```diff
@@ -71,6 +75,311 @@ def expand_dims(a, axis):
+    (array([[4, 3]], array([[0, 1]]))
+
+    >>> ht.sort(x, descending=True)
+    (array([[4, 1]], array([[0, 1]]))
+    (array([[3, 2]], array([[1, 0]]))
+    """
+    # default: using last axis
+    if axis is None:
+        axis = len(a.shape) - 1
+
+    stride_tricks.sanitize_axis(a.shape, axis)
+
+    if a.split is None or axis != a.split:
+        # sorting is not affected by split -> we can just sort along the axis
+        final_result, final_indices = torch.sort(a._DNDarray__array, dim=axis, descending=descending)
+
+    else:
+        # sorting is affected by split, processes need to communicate results
+        # transpose so we can work along the 0 axis
+        transposed = a._DNDarray__array.transpose(axis, 0)
+        local_sorted, local_indices = torch.sort(transposed, dim=0, descending=descending)
+
+        size = a.comm.Get_size()
+        rank = a.comm.Get_rank()
+        counts, disp, _ = a.comm.counts_displs_shape(a.gshape, axis=axis)
+
+        actual_indices = local_indices.to(dtype=local_sorted.dtype) + disp[rank]
+
+        length = local_sorted.size()[0]
+
+        # Separate the sorted tensor into size + 1 equal length partitions
+        partitions = [x * length // (size + 1) for x in range(1, size + 1)]
+        local_pivots = local_sorted[partitions] if counts[rank] else torch.empty(
+            (0, ) + local_sorted.size()[1:], dtype=local_sorted.dtype)
+
+        # Only processes with elements should share their pivots
+        gather_counts = [int(x > 0) * size for x in counts]
+        gather_displs = (0, ) + tuple(np.cumsum(gather_counts[:-1]))
+
+        pivot_dim = list(transposed.size())
+        pivot_dim[0] = size * sum([1 for x in counts if x > 0])
+
+        # share the local pivots with root process
+        pivot_buffer = torch.empty(pivot_dim, dtype=a.dtype.torch_type())
+        a.comm.Gatherv(local_pivots, (pivot_buffer, gather_counts, gather_displs), root=0)
+
+        pivot_dim[0] = size - 1
+        global_pivots = torch.empty(pivot_dim, dtype=a.dtype.torch_type())
+
+        # root process creates new pivots and shares them with other processes
+        if rank is 0:
+            sorted_pivots, _ = torch.sort(pivot_buffer, descending=descending, dim=0)
+            length = sorted_pivots.size()[0]
+            global_partitions = [x * length // size for x in range(1, size)]
+            global_pivots = sorted_pivots[global_partitions]
+
+        a.comm.Bcast(global_pivots, root=0)
+
+        # Create matrix that holds information which process gets how many values at which position
+        zeroes_dim = (size, ) + transposed.size()[1:]
+        partition_matrix = torch.zeros(zeroes_dim, dtype=torch.int64)
+
+        # Create matrix that holds information, which value is shipped to which process
+        index_matrix = torch.empty_like(local_sorted, dtype=torch.int64)
+
+        # Iterate along the split axis which is now 0 due to transpose
+        for i, x in enumerate(local_sorted):
```

Comment:
Question without having thought it through entirely. Is this loop and numpy conversion necessary here? Generally, this will add quite some overhead due to cpu-gpu transfer and due to Python ineffcient loops